### PR TITLE
Replace AggregatingProblemConsumer with ProblemSummarizer

### PR DESCRIPTION
**Pull request by reinsch82**
Fixes https://github.com/gradle/gradle/issues/30570

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test aCTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test ACTR

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/NoOpProblemSummarizer.java [hunk 1]
```diff
@@ -14,22 +14,16 @@
...
-import org.gradle.api.NonNullApi;
-import org.gradle.tooling.internal.protocol.problem.InternalDocumentationLink;
-import org.gradle.tooling.internal.protocol.problem.InternalSeverity;
+import org.gradle.internal.operations.OperationIdentifier;
 
 import javax.annotation.Nullable;
-import java.util.List;
 
-@NonNullApi
-public interface InternalProblemAggregationVersion2 extends InternalProblemAggregation {
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/NoOpProblemSummarizer.java:27]: ❌ This seems wrong. Removing protocol interfaces can potentially break cross-version compatibility.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummarizer.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.api.problems.internal;
+
+import org.gradle.internal.operations.OperationIdentifier;
+
+import javax.annotation.Nullable;
+
+public interface ProblemSummarizer {
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummarizer.java:23]: ❌ I remember we discussed that we'll introduce the summarization via delegation. Instead of a new interface, we should have a `ThresholdAwareProblemEmitter implements ProblemEmitter` that delegates to the existing emitter if the threshold value hasn't been reached.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemEmitter.java [hunk 3]
```diff
@@ -34,5 +34,4 @@ public interface ProblemEmitter {
      * @param problem The problem to emit.
      */
     void emit(Problem problem, @Nullable OperationIdentifier id);
-
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemEmitter.java:37]: 💄 unnecessary change

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.events.problems;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A problem summary
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java:22]: 💄 Missing . at the end of the sentence.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ *
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java:23]: 💄 Add `@see` annotation to ProblemSummaries.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummaries.java [hunk 6]
```diff
@@ -14,17 +14,24 @@
...
+import org.gradle.api.Incubating;
 
 import java.util.List;
 
-@NonNullApi
-public interface InternalProblemAggregationVersion3 extends InternalProblemDetailsVersion2 {
-
-    InternalProblemDefinition getProblemDefinition();
+/**
+ * All problems summaries
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummaries.java:24]: 💄 Missing . at the end of the sentence. Add @see annotation to ProblemsSummariesEvent.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.events.problems;
+
+import org.gradle.api.Incubating;
+
+/**
+ * The Event capturing all problems summaries
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java:22]: ❌ Missing javadoc:
- explain when the client can expect this event.
- and how the content should be interpreted.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java [hunk 8]
```diff
@@ -83,7 +88,7 @@ public void execute(WorkerProcessContext workerProcessContext) {
                     // Make the argument serializers available so work implementations can register their own serializers
                     registration.add(RequestArgumentSerializers.class, argumentSerializers);
                     registration.add(InstantiatorFactory.class, instantiatorFactory);
-                    registration.add(InternalProblems.class, new DefaultProblems(ImmutableList.of(new WorkerProblemEmitter(responder)), CurrentBuildOperationRef.instance()));
+                    registration.add(InternalProblems.class, new DefaultProblems(new SummarizerToEmitterAdapter(new WorkerProblemEmitter(responder)), CurrentBuildOperationRef.instance()));
```
Comment (donat) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java:91]: I don't understand why we need this.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemSummary.java [hunk 9]
```diff
@@ -18,7 +18,11 @@
 
 import org.gradle.api.NonNullApi;
 
+import java.io.Serializable;
+
 @NonNullApi
-public interface InternalProblemAggregationDetails extends InternalProblemDetails {
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemSummary.java:22]: ❌ Another change that may break cross-version compatibility.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemSummariesDetails.java [hunk 10]
```diff
@@ -17,13 +17,11 @@
 package org.gradle.tooling.internal.protocol;
 
 import org.gradle.api.NonNullApi;
-import org.gradle.tooling.internal.protocol.problem.InternalLabel;
-import org.gradle.tooling.internal.protocol.problem.InternalProblemCategory;
+import org.gradle.tooling.internal.protocol.problem.InternalProblemDetailsVersion2;
 
-@NonNullApi
-public interface InternalProblemAggregation {
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemSummariesDetails.java:24]: ❌ Potential breaking change.

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        problemSummariesEvent != null
+
+        def summaries = problemSummariesEvent.problemsSummaries.problemsSummaries
+        summaries.size() == 0
+
+        where:
+        exceedingCount << [-5, -1, 0]
+    }
+
+    boolean validateProblems(int totalSentEventsCount, List<SingleProblemEvent> problems) {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:92]: 💄 Move to below the test methods.

Comment (donat) [reply platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:92]: `validateFirstNProblems()` maybe?

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+                .addProgressListener(listener)
+                .forTasks("reportProblem")
+                .run()
+        }
+
+        then:
+        def problems = listener.problems
+        problems.size() == THRESHOLD_DEFAULT_VALUE + differentProblemCount
+        validateProblems(THRESHOLD_DEFAULT_VALUE, problems)
+
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:144]:
```suggestion
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        def problemSummariesEvent = listener.summariesEvent as ProblemsSummariesEvent
+
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:146]:
```suggestion
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+                .addProgressListener(listener)
+                .forTasks("reportProblem")
+                .run()
+        }
+
+        then:
+        def problems = listener.problems
+        problems.size() == thresholdInOption
+        validateProblems(thresholdInOption, problems)
+
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:174]:
```suggestion
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        def problemSummariesEvent = listener.summariesEvent as ProblemsSummariesEvent
+
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:176]:
```suggestion
```

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+ * @since 8.12
+ */
+@Incubating
+public interface ProblemsSummariesEvent extends ProblemEvent {
+    /**
+     * Get problems summaries
+     *
+     * @since 8.12
+     */
+    ProblemsSummaries getProblemsSummaries();
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java:33]: ❌ I think we can delete `ProblemsSummaries` and return `List<ProblemSummary>` here.

File: platforms/ide/problems-api/build.gradle.kts [hunk 17]
```diff
@@ -33,6 +33,7 @@ dependencies {
     api(projects.stdlibJavaExtensions)
     api(projects.baseServices)
     api(projects.buildOperations)
+    api(projects.buildOption)
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/ide/problems-api/build.gradle.kts:36]: The `DefaultProblemSummarizer` uses it to get the internal setting for the threshold.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemId.java [hunk 18]
```diff
@@ -83,6 +83,9 @@ public boolean equals(Object o) {
     public int hashCode() {
         int result = id.hashCode();
         result = 31 * result + parent.hashCode();
+        if (displayName != null) {
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemId.java:86]: I think there were NPE's after some of my changes.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java [hunk 19]
```diff
@@ -94,14 +95,11 @@ public Problem create(Action<InternalProblemSpec> action) {
      */
     @Override
     public void report(Problem problem) {
-        Throwable exception = problem.getException();
-        if(exception != null) {
-            problems.put(exception, problem);
-        }
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java:100]: This was for the failure association.

I had some warnings after moving things around. turns out removing the unneeded code was the best solution to the problem :)

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemSummarizer.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DefaultProblemSummarizer implements ProblemReporter, ProblemSummarizer {
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemSummarizer.java:38]: The `DefaultProblemSummarizer`has to create the final report which is facilitated by `org.gradle.problems.buildtree.ProblemReporter`. 
see the `report` method.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/NoOpProblemSummarizer.java (hunk 1 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/NoOpProblemSummarizer.java]: I think I did this before talking about the TAPI in depth.

File: platforms/ide/problems-api/build.gradle.kts (hunk 17 above)
[the remark this answers is not in the document]
Comment (donat) [reply platforms/ide/problems-api/build.gradle.kts:36]: I expected an `implementation` dependency for that.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummarizer.java (hunk 2 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummarizer.java:23]: That is correct.
I saw some issues with that approach. There is a list of `ProblemEmitter`s used to distribute to the event processing and the problems report. When the `ProblemsSummarizer` also  implements `ProblemsEmitter` it would also end up in that list. So I opted for a separate Interface to be able to guide the injection more precisely.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java (hunk 8 above)
Comment (reinsch82) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java:91]: related to this.

https://github.com/gradle/gradle/pull/30841/files/8ed7c2276eb2d8c4dc8d13c7e724844e84256ccb#r1827908749

File: platforms/ide/problems-api/build.gradle.kts (hunk 17 above)
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/ide/problems-api/build.gradle.kts:36]: the build complained about implementation.

I guess because I use the InternalOption instance in cross version tests.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemEmitter.java (hunk 3 above)
Comment (donat) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemEmitter.java:37]: There's still an unnecessary extra whitespace there.

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+        problemSummariesEvent != null
+        def summaries = problemSummariesEvent.problemsSummaries
+        summaries.size() == 0
+
+        where:
+        exceedingCount << [-5, -1, 0]
+    }
+
+    @TargetGradleVersion(">=8.10.2 <8.11")
+    def "No summaries old gradle version before 8.12"() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:91]: 💄 
```suggestion
    def "No summaries received from Gradle versions before 8.12"() {
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+        then:
+        def problems = listener.problems
+        problems.size() == THRESHOLD_DEFAULT_VALUE + differentProblemCount
+        validateFirstNProblems(THRESHOLD_DEFAULT_VALUE, problems)
+        def problemSummariesEvent = listener.summariesEvent as ProblemsSummariesEvent
+        def summaries = problemSummariesEvent.problemsSummaries
+        summaries.size() == 1
+    }
+
+    def "problem summarization threshold can be set by an internal option"() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:140]:
```suggestion
    def "Problem summarization threshold can be set by an internal option"() {
```

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java (hunk 7 above)
Comment (donat) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java:24]: This is still relevant.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.gradle.tooling.events.problems;
+
+import org.gradle.api.Incubating;
+
+import java.util.List;
+
+/**
+ * The Event capturing all problems summaries
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemsSummariesEvent.java:24]:
```suggestion
 * The event capturing all problems summaries.
```

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemAggregationDetails.java [hunk 24]
```diff
@@ -18,6 +18,10 @@
 
 import org.gradle.api.NonNullApi;
 
+/**
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemAggregationDetails.java:21]: For future reference, I think we should add a comment here, that we kept the interfaces for maintaining cross-version compatibility with TAPI version X where these classes are still being used.

(true for all modified classes in the protocol package)

**Review by donat — CHANGES_REQUESTED** [review]
Almost good to go. I have a few 💄 suggestions and one mandatory request to improve the javadoc on `ProblemsSummariesEvent`.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java (hunk 8 above)
Comment (donat) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java:91]: The url does not work.

**Review by hegyibalint — CHANGES_REQUESTED** [review]

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummarizer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummarizer.java:23]: Answer: to provide strict ordering, as the summarizer needs to run first

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemAggregationDetails.java [hunk 25]
```diff
@@ -18,6 +18,10 @@
...
+ * <p>DO NOT CHANGE THIS INTERFACE - it is part of the cross-version protocol.
+ */
+
```
Comment (hegyibalint) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemAggregationDetails.java:24]: 💄 why the newline?

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemSummary.java (hunk 9 above)
Comment (hegyibalint) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalProblemSummary.java:22]: Can you explain a bit more?

**Review by hegyibalint — APPROVED** [review]
We had a pairing session, approved after discussion

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package org.gradle.tooling.events.problems;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A problem summary.
+ *
+ * @see ProblemSummary
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java:24]: ❌ Self reference

**Review by donat — CHANGES_REQUESTED** [review]
One nitpick and it's good to go from my side.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java (hunk 26 above)
Comment (reinsch82) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummary.java]: embarrassing
